Display corrections in context menu for #54

speter52 opened the pull request:
Displaying the corrections for a misspelled word as context menu options as described in #54. This is my first OSS contribution so any feedback on the design or workflow would be great.

50Wliu commented on lib/spell-check-view.coffee:
```diff
@@ -107,3 +111,76 @@ class SpellCheckView
+  addContextMenuEntries: (mouseEvent) =>
+    @clearContextMenuEntries()
+    # Check to see if a context menu was opened through a right mouse click.
+    if mouseEvent.button is 2
```
Why do we care how the menu was opened?

50Wliu commented on lib/spell-check-view.coffee:
```diff
@@ -107,3 +111,76 @@ class SpellCheckView
     instance.suggest args, misspelling
+
+  initializeContextMenuListeners: =>
+    atom.views.getView(@editor).addEventListener 'mousedown', @addContextMenuEntries
```
You'll need to find a better way to detect when a context menu is opened - there are ways to open it without using the mouse (for example, Windows has a dedicated keyboard key that opens the menu).

speter52 replied:
Ah right. I assumed that context menus were only opened with right-mouse clicks so thought a `mousedown` listener would work. I'll look into a more reliable way of detecting when context menus open/close.

speter52 commented:
I found a `contextmenu` event in the Electron code which fires when context menus open, so I updated the listener to watch for that. I also modified the `SpellCheckView` class to hold a reference to a list of all context menu items that are added to the global registry - which is shared between the `SpellCheckView`'s of other editor instances. This way the menu items can be disposed as needed. I had some difficulty properly sharing that list between different `SpellCheckView`'s which I elaborated on at line 64 in `lib/main.coffee`.

Also the Travis CI check seems to be taking a lot longer than before?

speter52 commented:
Thanks, what would the next steps be?

50Wliu commented:
For someone with proper experience with the spell-check package to give this a review.  I believe that would be @as-cii.

as-cii commented:
This looks awesome, @speter52! I found a little quirk when testing this out: when a word has no corrections, `No corrections` is shown as expected. When closing the context menu and reopening it, `No corrections` is shown without a separator, and never goes away from there on.

![context-menu](https://cloud.githubusercontent.com/assets/482957/20381329/4c3cd3b2-aca7-11e6-8013-1a100e8a5892.gif)

Honestly I think we might get rid of the `No corrections` menu item, as it doesn't perform any action when clicked.

Other than this little nuance, I think this is ready to 🚢, so I will hit the merge button after addressing the minor problem described above. Nice work @speter52! 💯

speter52 commented:
Thanks a lot @as-cii! I just fixed the 'No corrections' issue. I had missed that use case when refactoring how the TextEditor's stored the context menu entries. Let me know if you see anything else needed from my end.

50Wliu commented:
I agree that we should get rid of the "No corrections" menu, if just to stay consistent with how Chrome does it.

as-cii commented on lib/spell-check-view.coffee:
```diff
@@ -107,3 +110,64 @@ class SpellCheckView
+      # Update the buffer with the correction.
+      @editor.setSelectedBufferRange(marker.getBufferRange())
+      @editor.insertText(correction.suggestion)
+    else
```
Do we need to handle the `else` scenario? If so, can we add a test for it?

speter52 replied:
I had brought this use case over from the `CorrectionsView:confirm()` function to preserve the 'Add to Known Words' functionality, it looks like that's the only way it's being used right now. I added a spec now to test adding a known word from the context menu.

as-cii commented:
I made a couple of changes and left an additional comment on a line of code that I missed during the first review. Other than that, this looks ready to merge! Again, great work! 💯

speter52 commented:
Just pushed the review changes!

as-cii commented:
Awesome, thanks @speter52! 💥

cameronmcefee commented:
> This is my first OSS contribution so any feedback on the design or workflow would be great.

👏 👏 👏  You're off to a great start. Thanks so much for this.

jbjonesjr commented:
:+1: this is great news. Would use your contributions again

colinbrislawn commented:
Well done! This program has needed this feature for ages!

speter52 commented:
Thanks for the comments, happy to contribute! 😀
